objects/src/accounts/vault.rs

```diff
@@ -17,37 +19,46 @@ use core::default::Default;
 /// An account vault can be reduced to a single hash which is the root of the Sparse Merkle tree.
 #[derive(Debug, Clone, Default, PartialEq, Eq)]
 pub struct AccountVault {
-    // TODO: add backing sparse Merkle tree
+    asset_tree: TieredSmt,
+    // TODO: Do we need this?  We could expose `.values()` on TieredSmt which allow us to return
+    // an iterator over all assets in the tree.  This makes the vault inefficient as it requires
+    // linear search for updates.  Alternatively we could use a HashMap / BTreeMap.
     assets: Vec<Asset>,
```

Comment:
I'm not sure this is needed? @bobbinth

Reply:
Agreed - I think we can get away just `TieredSmt`. So, it would look something like this:
```Rust
pub struct AccountVault {
    assets: TieredSmt,
}
```
And we can definitely expose `TieredSmt.values` - though, depending on what we want to return, we could use better naming (i.e., if we are just return an iterator over values, then `values()` is fine; but if we are returning an iterator over key-value pairs, maybe `entries()` would be better).

I didn't quite understand why it would that require linear search for updates (or did you mean if we keep `assets` as a vector here, then it would be linear?).

Reply:
One open question: what do we use for keys/values in the SMT.

For fungible assets, the key could be just `[0, 0, 0, faucet_id]` (we can't put amount into the key as we want to keep the key value independent of the amount). For value, we could have either:
- Full asset: `[balance, 0, 0, 0, faucet_id]` or
- Just the balance: `[balance, 0, 0, 0]`.

Probably storing the full asset makes more sense.

For non-fungible assets, the key would probably be the hash of the asset itself. And value would be the same as the key.

I do wonder if there is a more efficient construction we could come up with. For example, for fungible assets, we could use a simplified version of `TieredSmt` which works with single-element keys (and use `faucet_id` as the key).. But this won't work for non-fungible assets. We could also split the vault into two "sub-vaults": one for fungible assets and the other for non-fungible assets. May be worth it as a future optimization.

Reply:
For the purpose of the `AssetVault` I think it should be sufficient to just expose `values()` as the values are fully descriptive (the keys are some derivative of the value).  However, thinking more generally about the TSMT API I think both `values()` and `entries()` could be useful. 

I think storing the full asset (`[balance, 0, 0, 0, faucet_id]`) is preferable as it means that the TSMT values fully describe the assets in the vault without any reliance on the keys.

I think we should definitely consider it but it would be good to understand what you're thinking of in terms the `TieredSmt`. Could you elaborate on how you're thinking about the simplified version of the `TieredSmt` working?

Reply:
> Could you elaborate on how you're thinking about the simplified version of the `TieredSmt` working?

The simplified version would use a single elements for keys, but would still have tiers. So, the interface would be similar to `SimpleSmt` - but we'd get the benefit of compaction (i.e., shorter Merkle paths in most cases), more efficient in-memory representation, and also more efficient MASM code.

We could also use the same tree for the account DB as there we map 64-bit keys to hashes.

Reply:
Got it so the data structure would be the same but the interface would be constrained to a single `Felt`, makes sense to me.  I don't think it's an urgent requirement but would be nice to have.  We currently spend a few cycles creating the key:

https://github.com/0xPolygonMiden/miden-base/blob/a8e39271150ac6af7cb2eb7d9f798b98490e61f5/miden-lib/asm/sat/asset_vault.masm#L24-L26

Reply:
The biggest efficiency gain would not not here but in the data structure itself (in Rust implementation of it). Would probably reduce memory requirements quite a bit.